File: k9mail/src/main/java/com/fsck/k9/notification/NewMailNotifications.java
package com.fsck.k9.notification;


import android.app.Notification;
import android.support.v4.app.NotificationManagerCompat;
import android.util.SparseArray;

import com.fsck.k9.Account;
import com.fsck.k9.K9;
import com.fsck.k9.K9.NotificationHideSubject;
import com.fsck.k9.activity.MessageReference;
import com.fsck.k9.mailstore.LocalMessage;


/**
 * Handle notifications for new messages.
 * <p>
 * We call the notification shown on the device <em>summary notification</em>, even when there's only one new message.
 * Notifications on an Android Wear device are displayed as a stack of cards and that's why we call them <em>stacked
 * notifications</em>. We have to keep track of stacked notifications individually and recreate/update the summary
 * notification when one or more of the stacked notifications are added/removed.<br>
 * {@link NotificationData} keeps track of all data required to (re)create the actual system notifications.
 * </p>
 */
class NewMailNotifications {
    private final NotificationController controller;
    private final NotificationContentCreator contentCreator;
    private final DeviceNotifications deviceNotifications;
    private final WearNotifications wearNotifications;
    private final SparseArray<NotificationData> notifications = new SparseArray<NotificationData>();
    private final Object lock = new Object();


    NewMailNotifications(NotificationController controller, NotificationContentCreator contentCreator,
            DeviceNotifications deviceNotifications, WearNotifications wearNotifications) {
        this.controller = controller;
        this.deviceNotifications = deviceNotifications;
        this.wearNotifications = wearNotifications;
        this.contentCreator = contentCreator;
    }

    public static NewMailNotifications newInstance(NotificationController controller,
            NotificationActionCreator actionCreator) {
        NotificationContentCreator contentCreator = new NotificationContentCreator(controller.getContext());
        WearNotifications wearNotifications = new WearNotifications(controller, actionCreator);
        DeviceNotifications deviceNotifications = DeviceNotifications.newInstance(
                controller, actionCreator, wearNotifications);
        return new NewMailNotifications(controller, contentCreator, deviceNotifications, wearNotifications);
    }

    public void addNewMailNotification(Account account, LocalMessage message, int unreadMessageCount) {
        NotificationContent content = contentCreator.createFromMessage(account, message);

        synchronized (lock) {
            NotificationData notificationData = getOrCreateNotificationData(account, unreadMessageCount);
            AddNotificationResult result = notificationData.addNotificationContent(content);

            if (result.shouldCancelNotification()) {
                int notificationId = result.getNotificationId();
                cancelNotification(notificationId);
            }

            createStackedNotification(account, result.getNotificationHolder());
            createSummaryNotification(account, notificationData, false);
        }
    }

    public void removeNewMailNotification(Account account, MessageReference messageReference) {
        synchronized (lock) {
            NotificationData notificationData = getNotificationData(account);
            if (notificationData == null) {
                return;
            }

            RemoveNotificationResult result = notificationData.removeNotificationForMessage(messageReference);
            if (result.isUnknownNotification()) {
                return;
            }

            cancelNotification(result.getNotificationId());

            if (result.shouldCreateNotification()) {
                createStackedNotification(account, result.getNotificationHolder());
            }

            updateSummaryNotification(account, notificationData);
        }
    }

    public void clearNewMailNotifications(Account account) {
        NotificationData notificationData;
        synchronized (lock) {
            notificationData = removeNotificationData(account);
        }

        if (notificationData == null) {
            return;
        }

        for (int notificationId : notificationData.getActiveNotificationIds()) {
            cancelNotification(notificationId);
        }

        int notificationId = NotificationIds.getNewMailSummaryNotificationId(account);
        cancelNotification(notificationId);
    }

    private NotificationData getOrCreateNotificationData(Account account, int unreadMessageCount) {
        NotificationData notificationData = getNotificationData(account);
        if (notificationData != null) {
            return notificationData;
        }

        int accountNumber = account.getAccountNumber();
        NotificationData newNotificationHolder = createNotificationData(account, unreadMessageCount);
        notifications.put(accountNumber, newNotificationHolder);

        return newNotificationHolder;
    }

    private NotificationData getNotificationData(Account account) {
        int accountNumber = account.getAccountNumber();
        return notifications.get(accountNumber);
    }

    private NotificationData removeNotificationData(Account account) {
        int accountNumber = account.getAccountNumber();
        NotificationData notificationData = notifications.get(accountNumber);
        notifications.remove(accountNumber);
        return notificationData;
    }

    NotificationData createNotificationData(Account account, int unreadMessageCount) {
        NotificationData notificationData = new NotificationData(account);
        notificationData.setUnreadMessageCount(unreadMessageCount);
        return notificationData;
    }

    private void cancelNotification(int notificationId) {
        getNotificationManager().cancel(notificationId);
    }

    private void updateSummaryNotification(Account account, NotificationData notificationData) {
        if (notificationData.getNewMessagesCount() == 0) {
            clearNewMailNotifications(account);
        } else {
            createSummaryNotification(account, notificationData, true);
        }
    }

    private void createSummaryNotification(Account account, NotificationData notificationData, boolean silent) {
        Notification notification = deviceNotifications.buildSummaryNotification(account, notificationData, silent);
        int notificationId = NotificationIds.getNewMailSummaryNotificationId(account);

        getNotificationManager().notify(notificationId, notification);
    }

    private void createStackedNotification(Account account, NotificationHolder holder) {
        if (isPrivacyModeEnabled()) {
            return;
        }

        Notification notification = wearNotifications.buildStackedNotification(account, holder);
        int notificationId = holder.notificationId;

        getNotificationManager().notify(notificationId, notification);
    }

    private boolean isPrivacyModeEnabled() {
        return K9.getNotificationHideSubject() != NotificationHideSubject.NEVER;
    }

    private NotificationManagerCompat getNotificationManager() {
        return controller.getNotificationManager();
    }
}

File: app/core/src/main/java/com/fsck/k9/notification/NewMailNotifications.kt
package com.fsck.k9.notification

import android.util.SparseArray
import androidx.core.app.NotificationManagerCompat
import com.fsck.k9.Account
import com.fsck.k9.K9
import com.fsck.k9.controller.MessageReference
import com.fsck.k9.mailstore.LocalMessage

/**
 * Handle notifications for new messages.
 */
internal open class NewMailNotifications(
    private val notificationHelper: NotificationHelper,
    private val contentCreator: NotificationContentCreator,
    private val messageSummaryNotifications: MessageSummaryNotifications,
    private val singleMessageNotifications: SingleMessageNotifications
) {
    private val notifications = SparseArray<NotificationData>()
    private val lock = Any()

    fun addNewMailNotification(account: Account, message: LocalMessage, unreadMessageCount: Int, silent: Boolean) {
        val content = contentCreator.createFromMessage(account, message)

        synchronized(lock) {
            val notificationData = getOrCreateNotificationData(account, unreadMessageCount)

            val result = notificationData.addNotificationContent(content)
            if (result.shouldCancelNotification) {
                val notificationId = result.notificationId
                cancelNotification(notificationId)
            }

            createSingleMessageNotification(account, result.notificationHolder)
            createSummaryNotification(account, notificationData, silent)
        }
    }

    fun removeNewMailNotification(account: Account, messageReference: MessageReference) {
        synchronized(lock) {
            val notificationData = getNotificationData(account) ?: return

            val result = notificationData.removeNotificationForMessage(messageReference)
            if (result.isUnknownNotification) return

            cancelNotification(result.notificationId)

            if (result.shouldCreateNotification) {
                createSingleMessageNotification(account, result.notificationHolder)
            }

            updateSummaryNotification(account, notificationData)
        }
    }

    fun clearNewMailNotifications(account: Account) {
        val notificationData = synchronized(lock) { removeNotificationData(account) } ?: return

        for (notificationId in notificationData.getActiveNotificationIds()) {
            cancelNotification(notificationId)
        }

        val notificationId = NotificationIds.getNewMailSummaryNotificationId(account)
        cancelNotification(notificationId)
    }

    private fun getOrCreateNotificationData(account: Account, unreadMessageCount: Int): NotificationData {
        val notificationData = getNotificationData(account)
        if (notificationData != null) return notificationData

        val accountNumber = account.accountNumber
        val newNotificationHolder = createNotificationData(account, unreadMessageCount)
        notifications.put(accountNumber, newNotificationHolder)

        return newNotificationHolder
    }

    private fun getNotificationData(account: Account): NotificationData? {
        val accountNumber = account.accountNumber
        return notifications[accountNumber]
    }

    private fun removeNotificationData(account: Account): NotificationData? {
        val accountNumber = account.accountNumber
        val notificationData = notifications[accountNumber]
        notifications.remove(accountNumber)
        return notificationData
    }

    protected open fun createNotificationData(account: Account, unreadMessageCount: Int): NotificationData {
        return NotificationData(account, unreadMessageCount)
    }

    private fun cancelNotification(notificationId: Int) {
        notificationManager.cancel(notificationId)
    }

    private fun updateSummaryNotification(account: Account, notificationData: NotificationData) {
        if (notificationData.newMessagesCount == 0) {
            clearNewMailNotifications(account)
        } else {
            createSummaryNotification(account, notificationData, silent = true)
        }
    }

    private fun createSummaryNotification(account: Account, notificationData: NotificationData, silent: Boolean) {
        val notification = messageSummaryNotifications.buildSummaryNotification(account, notificationData, silent)
        val notificationId = NotificationIds.getNewMailSummaryNotificationId(account)
        notificationManager.notify(notificationId, notification)
    }

    private fun createSingleMessageNotification(account: Account, holder: NotificationHolder) {
        if (isPrivacyModeEnabled) {
            return
        }

        val notification = singleMessageNotifications.buildSingleMessageNotification(account, holder)
        val notificationId = holder.notificationId
        notificationManager.notify(notificationId, notification)
    }

    private val isPrivacyModeEnabled: Boolean
        get() = K9.notificationHideSubject !== K9.NotificationHideSubject.NEVER

    private val notificationManager: NotificationManagerCompat
        get() = notificationHelper.getNotificationManager()
}
